King rework - revived

Lumipharon opened the pull request:
## About The Pull Request
It's back baby, new and improved
![image](https://user-images.githubusercontent.com/7869430/209131524-78de795b-08fa-44de-8659-c0edbf82a9db.png)
_(picture may not be indicative of final product)_


Revived after Tivi died to death.
A couple of minor tweaks from the original PR. 
- Petrify has a slightly higher range and shorter windup.
- Offguard causes confusion as well as scatterloss
- Zero form beam now critically uses the unused king screech sprite
- Zero form goes onto 5 second cooldown if interrupted while charging (no more spamming sounds)

Also I found that xeno icons were (as far as I can tell) needlessly updated every life tick. I removed this and added it to movement intent switching, which was the only time it appeared to be relevant.

Also added a new ability: Shattering roar
After a one second windup, unleashes a wave that briefly knocks down anyone in it's path, causing some minor damage and a bit of staggerslow, that scales with distance to the king. Also shatters windows for dramatic effect.

https://user-images.githubusercontent.com/7869430/209316369-27cc3690-6963-416b-87e1-986b01e57df5.mp4



There wasn't a huge amount of feedback on the original PR, so please give feedback here or in the feedback thread on discord, as I'm very open to changing this a lot - there's some stuff I'm already considering, but Tivi had it fully functional and far more interesting than current king, so I'd also be just as happy to merge this and continue adjusting later.

From the original PR #11610 
REMOVED
Grav crush- has been nothing but issues due to it using bomb armor and it wasnt an interesting ability tbh

ADDED
Petrify - turns humans in view to stone for a set amount of time after a windup. petrified humans cannot be attacked
Off-guard - gives humans a debuff that drastically increases their scatter for a duration. effect decreases over time.
Zero-form beam: new primordial ability for king, fires a beam for 10 tiles that goes through walls and hurts humans that stand in it, has a windup

CHANGED
King can no longer be purchased and is available like queen, except it needs 15 xenos instead
Has same playtime requirements as queen

## Why It's Good For The Game

Tries to move king from its old and cost-decrease crept role into a more xeno support role
## Changelog
:cl: Code: Tivi & Lumi Sprites: Ghost
add: reworked king: removed grav crush, added 4 new abilities and made it available through evo instead of blessings
fix: Godmode prevents being hugged, stung, or breathing in smoke
/:cl:

Nicolai904 commented:
nice changes the shorter windup on petrify was really needed also sometiems marine can still shoot while petrified i am not sure how to replicate this bug however

novaepee commented:
you drop this, king
![image](https://user-images.githubusercontent.com/6610922/209210236-d4d4f38d-8840-4395-914d-6fc63d976d92.png)

Edelor68 commented:
Can marines just not be attacked while petrified, or are they immune to all damage?

GreatKingWombat commented:
> Can marines just not be attacked while petrified, or are they immune to all damage?

It adds god mode, so looks like immunity. Will be fun to see if the ability can bug out and fail to remove god mode.

BraveMole commented on code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:
```diff
@@ -32,104 +32,359 @@
+	for(var/mob/living/carbon/human/human in view(PETRIFY_RANGE, owner.loc))
+		if(is_blind(human))
+			continue
+		humans += human
+		human.notransform = TRUE
+		human.status_flags |= GODMODE
+		human.add_atom_colour(COLOR_GRAY, TEMPORARY_COLOUR_PRIORITY)
+		ADD_TRAIT(human, TRAIT_HANDS_BLOCKED, REF(src))
+		human.move_resist = MOVE_FORCE_OVERPOWERING
+		human.log_message("has been petrified by [owner] for [PETRIFY_DURATION] ticks", LOG_ATTACK, color="pink")
+		var/image/stone_overlay = image('icons/effects/effects.dmi', null, "petrified_overlay")
+		stone_overlay.filters += filter(arglist(alpha_mask_filter(render_source="*[REF(human)]",flags=MASK_INVERSE)))
+		var/mutable_appearance/mask = mutable_appearance()
+		mask.appearance = human.appearance
+		mask.render_target = "*[REF(human)]"
+		mask.alpha = 125
+		stone_overlay.overlays += mask
+		human.overlays += stone_overlay
+		humans[human] = stone_overlay
```
Try to add some lines between logic blocks, so it's not a wall of text

Lumipharon replied:
Tidied it up.

BraveMole commented on code/modules/mob/living/carbon/xenomorph/evolution.dm:
```diff
@@ -259,6 +280,10 @@
 		if(length(hive.xenos_by_typepath[/mob/living/carbon/xenomorph/hivemind]))
 			to_chat(src, span_warning("There cannot be two manifestations of the hivemind's will at once."))
 			return
+	else if(new_caste_type == /mob/living/carbon/xenomorph/king)
```
could be one if

Lumipharon replied:
there's some other else stuff there so I'mma just leave it for now.

Aporhtonoma commented:
> Also added a new ability: Shattering roar
> After a one second windup, unleashes a wave that briefly knocks down anyone in it's path, causing some minor damage and a bit of staggerslow, that scales with distance to the king. Also shatters windows for dramatic effect.

The benos have a proper engange ability the kingus is no longer just a phero machine with funny beam ... good change :)

shelteredfreak commented:
Can you make Kangus slash damage more linear? Like from 20/20/25/30 to 24/26/28/30? (Yes I also commented this on the original PR, I still want it) Will give a comment for feedback once I get the chance to try this.

Lumipharon commented:
melee scaling is more linear, adds some missing flags and options (i.e. intelligent, blessings), fixed some hotkey issues for roar and offguard, and fixed some issues with icon updates.

XSlayer300 commented:
King roar should shake the screen and use the same shockwave effect from the explosions.

Lumipharon commented:
> King roar should shake the screen and use the same shockwave effect from the explosions, and King should play the roar sound **when** they actually roar.
> 
> Offguard is still a bit useless on melee and flame bois, but with confusion, at least it helps.
> 
> Also, forgor to mention is that when muhreens get hit by the funny laser, make them hit using a red overlay, like getting hit by a bullet.

I did actually mean to add screen shake, I think I just forgot.
Roar has two roar sounds. The first is to indicate that are trying to activate it, the second (longer) roar is when it successfully casts.

Lumipharon commented:
Added a stagger resistant trait (as opposed to stagger immune). Currently it just provided stagger immunity against projectiles like queen, but I may do some other stuff with it later.

King and queen both have stagger resistance, however king loses it when channeling any of its abilities (i.e. beam, roar and petrify), giving marines a brief window where they can stagger cancel the ability via a shotgun to the face or whatever.

I strongly preferred this to 24/7 projectile stagger resistance like queen, but it did need at least some level of protection, so I'm happy with this.

Lumipharon commented:
Some final small tweaks. 

- Some damage/ability feedback
- nightfall should now work on mechs
- offguard makes less particle spam
- Roar has 35 sec CD instead of 25
- Speed for elder and up decrease very slightly

MiniMeatwad commented:
CC from hell lvl. 100 mob boss instead of a cade buster? Sure, whatever floats your boat. I too enjoy power fantasy.

Lumipharon commented:
Can't stop tweaking until it's merged.

Added immobility when channeling petrify, roar and beam (just the initial channel of beam, not while actually blasting) to make it easier to punish bad positioning/greedy plays.

Lumipharon commented:
A few godmode oversights. Godmode now prevents getting hugged, getting stung, and breathing in smoke.

Lumipharon commented:
Zero form beam now hits mechs for fat damage. Considering they can't be stunned it would be quite hard to catch them in the beam for any period of time, but now they can't simply ignore.

iLikeMemes commented:
Do not forget to remove CASTE_CAN_BECOME_KING from each castedatum and anything related if this gets properly merged.

Losenis commented:
You can get help intent shuffled while petrified, xenos are starting to do that to move marines into chokes.

Waselon commented:
> Tries to move king from its old and cost-decrease crept role into a more xeno support role

Thats a xeno battletank, not a support rn. I think it might need just a little bit of damage and/or hp nerf. 700 health at ancient is a lot

TiviPlus commented on code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:
```diff
@@ -32,104 +32,420 @@
+
+	finish_charging()
+	playsound(owner, 'sound/voice/xenos_roaring.ogg', 100)
+	for(var/mob/living/carbon/human/human_victim AS in GLOB.alive_human_list)
+		if(get_dist(human_victim, owner) > 9)
+			continue
```
use by z lists

TiviPlus commented on code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:
```diff
@@ -32,104 +32,420 @@
+
+	add_cooldown()
+	succeed_activate()
+
+/datum/action/xeno_action/activable/shattering_roar/proc/execute_attack(iteration, list/turf/turfs_to_attack, range, target, turf/source)
```
autodoc

TiviPlus commented on code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:
```diff
@@ -32,104 +32,420 @@
+
+	iteration++
+	addtimer(CALLBACK(src, .proc/execute_attack, iteration, turfs_to_attack, range, target, source), SHATTERING_ROAR_SPEED)
+
+/datum/action/xeno_action/activable/shattering_roar/proc/attack_turf(turf/turf_victim, severity)
```
autodoc

TiviPlus reviewed (changes requested):
res lgtm
